### PR TITLE
Better C# property validation

### PR DESCRIPTION
**Pull request by externl**
Adds better property name validation like in #2838
Part of #2746

### CODE REVIEW

File: csharp/src/Ice/Internal/Property.cs [hunk 1]
```diff
@@ -1,5 +1,17 @@
 // Copyright (c) ZeroC, Inc.
 
+#nullable enable
+
 namespace Ice.Internal;
 
-public sealed record class Property(string pattern, bool usesRegex, string defaultValue, bool deprecated);
+public sealed record class PropertyArray(
```
Comment (externl) [inline csharp/src/Ice/Internal/Property.cs:7]: Added a new type to so we can hold the name like in C++.

File: csharp/src/Ice/Properties.cs [hunk 2]
```diff
@@ -514,36 +558,25 @@ public List<string> getUnusedProperties()
...
                 break;
             }
 
             // As a courtesy to the user, perform a case-insensitive match and suggest the correct property.
             // Otherwise no other warning is issued.
-            if (logWarnings && propPrefix.Equals(prefix, StringComparison.OrdinalIgnoreCase))
+            if (logWarnings && properties.name.Equals(prefix, StringComparison.OrdinalIgnoreCase))
             {
-                logger.warning("unknown property: `" + key + "'; did you mean `" + propPrefix + "'?");
+                logger.warning("unknown property: `" + key + "'; did you mean '" + properties.name + "'?");
```
[the remark this answers is not in the document]
Comment (externl) [reply csharp/src/Ice/Properties.cs:579]: No, I plan on fixing this in all languages later.

File: csharp/src/Ice/Internal/Property.cs [hunk 3]
```diff
@@ -1,5 +1,17 @@
 // Copyright (c) ZeroC, Inc.
 
+#nullable enable
+
 namespace Ice.Internal;
 
-public sealed record class Property(string pattern, bool usesRegex, string defaultValue, bool deprecated);
+internal sealed record class PropertyArray(
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/Property.cs:7]: should be on a single line.

File: csharp/src/Ice/Internal/PropertyNames.cs [hunk 4]
```diff
@@ -8,450 +8,230 @@ namespace Ice.Internal;
...
-    new(@"Ice.Warn.Datagrams", false, "0", false),
-    new(@"Ice.Warn.Dispatch", false, "1", false),
-    new(@"Ice.Warn.Endpoints", false, "1", false),
-    new(@"Ice.Warn.UnusedProperties", false, "0", false),
-    new(@"Ice.CacheMessageBuffers", false, "2", false),
-    };
+    internal static PropertyArray ProxyProps = new(
+        "Proxy",
+        [
+            new(@"EndpointSelection", false, "", false, null, false),
```
Comment (bernardnormier) [inline csharp/src/Ice/Internal/PropertyNames.cs:14]: I would use the parameter names for greater readability.

File: csharp/src/Ice/Properties.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/Properties.cs:579]: We prefer interpolated strings over "+" in C#.

**Review by bernardnormier — APPROVED** [review]

File: csharp/src/Ice/Properties.cs [hunk 5]
```diff
@@ -496,13 +496,57 @@ public List<string> getUnusedProperties()
...
+    /// <param name="properties">The properties to consider. </param>
+    /// <param name="propertyArray">The property array to search against.</param>
+    /// <exception cref="UnknownPropertyException"> Thrown if unknown properties are found.</exception>
+    internal static void validatePropertiesWithPrefix(string prefix, Properties properties, PropertyArray propertyArray)
+    {
+        // Do not check for unknown properties if Ice prefix, ie Ice, Glacier2, etc
+        foreach (string name in PropertyNames.clPropNames)
+        {
+            string icePrefix = name + ".";
+            if (prefix.StartsWith(icePrefix, StringComparison.Ordinal))
```
Comment (pepone) [inline csharp/src/Ice/Properties.cs:512]:
```suggestion
            string icePrefix = name + ".";
            if (prefix.StartsWith($"{name}.", StringComparison.Ordinal))
```

**Review by pepone — APPROVED** [review]

**Pull request merged**